profiling/src/profile/mod.rs

```diff
@@ -121,7 +125,7 @@ impl UpscalingRule {
 
 pub struct Profile {
     sample_types: Vec<ValueType>,
-    samples: FxIndexMap<Sample, Vec<i64>>,
+    samples: FxIndexMap<Sample, HashMap<Option<Timestamp>, Vec<i64>>>,
```

Comment:
There's a bit of a memory trade-off here: by making this more efficient for timeline data, we're creating an extra `HashMap` (and `Option`?) for every sample, even though there may not ever be any timestamp data.

Note tham I'm not sure this is a blocker -- maybe this is good enough for the other cases that we don't care that much, but I wanted to raise it so we know we're making some cases worse.

Reply:
The `Option` is over a `NonZeroi64` so in memory, this translates to `HashMap<i64, Vec<i64>>`.  
My expectation is that in the case where we don't have a timeline, we will have relitively few samples, so paying the cost of a small `HashMap` should be affordable.  One option would be to move to `BTreeMap`, which is slower but appears to have somewhat less (and more predictable) memory usage.

Reply:
On the ruby test, there is no significant difference with timeline off

With the `HashMap`
```
config: {:threads=>16, :depth=>50, :seconds=>60, :timeline_enabled=>false}
Before serialization, process rss usage is 64896k
After serialization, process rss usage is 64992k
Size of uncompressed serialized pprof: 3k
After shrinking, process rss usage is 64992k
sample_map false
```

Without
```
config: {:threads=>16, :depth=>50, :seconds=>60, :timeline_enabled=>false}
Before serialization, process rss usage is 64080k
After serialization, process rss usage is 64176k
Size of uncompressed serialized pprof: 3k
After shrinking, process rss usage is 64192k
main false
```

Reply:
The Ruby testing app is probably not a good proxy for this specific comparison, since the stacks never change, so you'll get a very low amount of samples.

I'm not sure if we have metrics for the average amount of stacks in a pprof in the backend, but if not, maybe choosing a few internal services and checking how many unique samples they have may provide a reasonable comparison point maybe?

Reply:
👋 I wondering why a `HashMap<Option<Timestamp>, Vec<i64>>>`  ? I was thinking about a `Vec<Pair<Timestamp, Vec<i64>>>` ? (sorry for my Rust, I do not know if `Pair` exists or is there an equivalent.

Because the probability that two identical samples have the same timestamp is really low, no?

Reply:
> Because the probability that two identical samples have the same timestamp is really low, no?

I actually discussed this with @danielsn at some point. I was thinking that in a normal situation yes, but if we are run in an OS or VM where the system timestamp counter is not extremely accurate, we may get different samples with the same timestamp.

Which doesn't mean that we have to handle them by aggregation -- we could as well return an error back at the caller. I do still think it's worth dealing with explicitly, vs not handling it very gracefully.